### PR TITLE
Fix syncer download order and add sync tests

### LINKED ISSUE #3114: Spawn the syncer as an async task
## Motivation

Zebra's syncer currently runs in the initial async task on the main thread. This means that it is launched and terminated differently to other tasks.

It also shares the thread with more urgent tasks, such as interrupts and signals. But that should be handled by #1351.

### Designs

- Make the `PeerSet` service sync and send (needs a Rust expert)
- Spawn the syncer in a separate task

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We want the syncer to download blocks in chain order, as documented.

We want to make sure that the syncer correctly drops blocks that are a long way ahead of the state tip.
This is a follow-up test for PR #3167 for ticket #1389.

## Solution

`ChainSync` bug fixes:
- Download synced blocks in chain order, not arbitrary `HashSet` order
    - This adds a direct dependency on `indexmap` (multiple Zebra dependencies already use `indexmap`)
- Increase the syncer lookahead height limit, because in-order syncs are more efficient
- Spawn the syncer into its own task - closes #3114

Extra Tests:
- Add `ChainSync` tests for `obtain_tips`, `extend_tips`, and wrong block hashes
- Add block too high tests for `obtain_tips` and `extend_tips`
- Add `ChainSync` tests for duplicate `FindBlocks` response hashes 

Make some futures `std::marker::Send` or `std::marker::Sync`, so they can be spawned: 
- Refactor so that `RetryLimit::Future` is `std::marker::Sync`
- Make the `ChainSync::sync` future `std::marker::Send` by spawning tips futures

Improve test harnesses:
- Improve `MockService` failure messages
- Add closure-based responses to the `MockService` API
- Move `MockChainTip` to `zebra-chain`
- Add a `MockChainTipSender` type alias
- Support `MockChainTip` in `ChainSync` and its downloader

## Review

I'm not sure who is around to review this PR, maybe @upbqdn?

### Reviewer Checklist

  - [x] Code implements Specs and Designs
  - [x] Tests for Expected Behaviour
  - [x] Tests for Errors

### CODE REVIEW

**Comment by teor2345** [comment]
> ```diff
> + Coverage   77.18%   77.90%   +0.72%     
> ```

This is a really nice test coverage increase.

File: zebra-chain/src/chain_tip/mock.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    }
+}
+
+impl ChainTip for MockChainTip {
+    fn best_tip_height(&self) -> Option<block::Height> {
+        *self.best_tip_height.borrow()
+    }
+
+    fn best_tip_hash(&self) -> Option<block::Hash> {
+        unreachable!("Method not used in `MinimumPeerVersion` tests");
```
Comment (teor2345) [inline zebra-chain/src/chain_tip/mock.rs]:
```suggestion
        unreachable!("Method not used in tests");
```

File: zebra-chain/src/chain_tip/mock.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    fn best_tip_mined_transaction_ids(&self) -> Arc<[transaction::Hash]> {
+        unreachable!("Method not used in `MinimumPeerVersion` tests");
```
Comment (teor2345) [inline zebra-chain/src/chain_tip/mock.rs]:
```suggestion
        unreachable!("Method not used in tests");
```

File: zebrad/src/components/sync.rs [hunk 3]
```diff
@@ -740,7 +782,7 @@ where
         }
     }
 
-    fn update_metrics(&self) {
+    fn update_metrics(&mut self) {
```
Comment (conradoplg) [inline zebrad/src/components/sync.rs:785]: Why is this needed?

File: zebrad/src/components/sync/downloads.rs [hunk 4]
```diff
@@ -395,7 +408,7 @@ where
     }
 
     /// Get the number of currently in-flight download tasks.
-    pub fn in_flight(&self) -> usize {
+    pub fn in_flight(&mut self) -> usize {
```
Comment (conradoplg) [inline zebrad/src/components/sync/downloads.rs:411]: Why is this needed?

**Review by conradoplg — APPROVED** [review]
Looks good! The tests are great.

There are two `mut`s that I don't understand why they're needed, but it's not a big deal. I'm approving this and we can take a look later if needed.

File: zebrad/src/components/sync.rs (hunk 3 above)
Comment (upbqdn) [reply zebrad/src/components/sync.rs:785]: If I understand it correctly, those `mut`s make the compiler aware that `update_metrics` has the sole access to `self`. This is needed in a multi-threaded environment, but I can't remember why. I had a discussion about this with Teor.
